File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    int count = 0;
    
    public int goodNodes(TreeNode root) {
        // approach: DFS while keeping track of max
        helper(root, Double.NEGATIVE_INFINITY);
        return count;
    }
    
    private void helper(TreeNode node, double max) {
        if (node == null) {
            return;
        }
        
        if (node.val >= max) {
            System.out.println(node.val + " >= " + max);
            count++;
        }
        
        max = Math.max(max, node.val);
        
        helper(node.left, max);
        helper(node.right, max);
    }
}

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        # approach: DFS, keep track of max
        
        def dfs(node, maximum):
            if not node:
                return 0
            
            if node.val < maximum:
                return dfs(node.left, maximum) + dfs(node.right, maximum)
            
            maximum = max(maximum, node.val)
            return 1 + dfs(node.left, maximum) + dfs(node.right, maximum)
        
        return dfs(root, float('-inf'))